### PR TITLE
User can create profile

### PR DESCRIPTION
**Pull request by zanenkn**
### Changes proposed in this pull request:

* Feature tests for happy and sad paths for the scenario
* Creates a Profile (host profile) model, columns in db: name, location, description, picture.
* Defines 'new' and 'create' methods in the Profiles controller
* Defines 'new' and 'create' as resources for Profile in the routes
* Implements Devise and refactors the navbar in order to only allow registered and logged in users to create a Host profile
* Merging with upstream development (hence the listing files appearing in 'Files changed')
* Authenticates the User for seeing all the views except index
* Restricts User from creating several profiles

### What we have learned working on this feature:

* refreshed on Devise helpers
* learned how to delete a column from the database via Rails console command
* learned about partials and how to work with those
* important lesson and insights about writing user stories (thanks Thomas!)
* `before_action` controller helper method

### Screenshots
image

### CODE REVIEW

File: app/views/partials/_form_profile.html.haml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  %p
+    = form.label :location
+    %br/
+    = form.text_field :location
+  %p
+    = form.label :description
+    %br/
+    = form.text_area :description  
+  %p
+    = form.label :price_kr_per_day
```
Comment (tochman) [inline app/views/partials/_form_profile.html.haml]:
```
= form.label :price, 'Price kr per day'
```

Comment (zanenkn) [reply app/views/partials/_form_profile.html.haml]: we remove the price altogether, that will now be an attribute of Offer model.

**Review by leiter007 — COMMENTED** [review]
I think it looks good with the Profile model, and the login functionality! All unit tests and the feature scenario seems reasonable. My only comment is that this PR strangely includes all the changes me and Carla did with the PR that was merged yesterday - connecting to "Visitor can create listing". This is why so many files seem to have been changed here right. Don't know why this is the case, I assume upstream development should recognise that this is not "new" code in that sense.. But if there aren't conflicts I guess it's fine!

**Comment by zanenkn** [comment]
yes, i think, this is because this PR was started yesterday and then we pulled upstream today. Although yes, it should not be recognised as "new" code. Coaches, do you have an input how to avoid this kind of stuff?

**Review by boamatule — APPROVED** [review]
Reviewed and it all seems to be pretty clear.

File: app/controllers/profiles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class ProfilesController < ApplicationController
+  
+  def new
+    @profile = Profile.new
+  end
+
+  def create
+    user = current_user
+    profile = Profile.create(profile_params)
+    user.profile_id = profile.id
```
Comment (tochman) [inline app/controllers/profiles_controller.rb]: Please consider this syntax:
```
current_user.profile.create(profile_params)
```
This, of course, assumes that you have a `has_one` association between `User` and `Profile`

File: app/controllers/profiles_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    
+    if profile.persisted?
+      user.save
```
Comment (tochman) [inline app/controllers/profiles_controller.rb]: Why do you save the user again?

File: app/models/profile.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class Profile < ApplicationRecord
+  validates_presence_of :name, :location, :description, :picture
+  has_one :user
```
Comment (tochman) [inline app/models/profile.rb]: Hmm... shouldn't this be `belongs_to`?

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  belongs_to :profile, optional: true
```
Comment (tochman) [inline app/models/user.rb]: Hmm... shouldn't this be `has_one`?

File: app/views/listings/new.html.haml [hunk 6]
```diff
@@ -1,3 +1,3 @@
 %h1 Create new listing
 
-= render "./partials/form.html.haml"
\ No newline at end of file
+= render "./partials/form_listing.html.haml"
```
Comment (tochman) [inline app/views/listings/new.html.haml]: Can you justify the name change? Why is this partial not in the `listings` folder?

File: db/migrate/20190509223219_add_profile_to_users.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddProfileToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :users, :profile, foreign_key: true
```
Comment (tochman) [inline db/migrate/20190509223219_add_profile_to_users.rb]: IMO, this should be:
```
add_reference :profiles, :users, foreign_key: true
```

File: app/controllers/profiles_controller.rb (hunk 2 above)
Comment (raptorf1) [reply app/controllers/profiles_controller.rb]: ok. done.

File: app/controllers/profiles_controller.rb (hunk 3 above)
Comment (raptorf1) [reply app/controllers/profiles_controller.rb]: According to our inspections with `pry` and `rails console`, If we do not have this line then the Profile is not passed in the User and we cannot verify further on whether the logged in user has a profile or not.
Let us say here that the tests are passing even without this line.

File: app/models/profile.rb (hunk 4 above)
Comment (raptorf1) [reply app/models/profile.rb]: Yesterday we tried the way you are suggest like that:
`rails g migration add_user_to_profiles user:references`
profile `belongs_to :user`
user `has_one :profile`
But then every profile had a `user_id` column. And then we had to call `profile.user` to get the user associated with the profile.
But we wanted to use the `user.profile` method in order to find if the logged in user already has a profile and hide the relevant link in the NavBar.

So we did:
`rails g migration add_profile_to_users profile:references`
profile `has one :user`
user `belongs to :profile`

We tried:
`rails g migration add_profile_to_users profile:references`
and then
profile `belongs_to :user`
user `has_one :profile`
but the tests fail.

So we believe we leave it as is. Unless you have a strong objection.

File: app/models/user.rb (hunk 5 above)
Comment (raptorf1) [reply app/models/user.rb]: Replied above

File: app/views/listings/new.html.haml (hunk 6 above)
Comment (raptorf1) [reply app/views/listings/new.html.haml]: We have a `partials` folder for all our partials.
As for the name change, we introduced another form partial (the profile one) so we wanted to differentiate between those two clearly (form_listing.html.haml was previously form.html.haml).

File: db/migrate/20190509223219_add_profile_to_users.rb (hunk 7 above)
Comment (raptorf1) [reply db/migrate/20190509223219_add_profile_to_users.rb]: Replied with our logic above. Waiting for guidance.

File: app/models/profile.rb (hunk 4 above)
Comment (tochman) [reply app/models/profile.rb]: IMO that association should be that a user has one profile and profile belongs to a user. For that, you need to add `user_id` field to the `profiles` table. The way you have it now does not make sense top me.

File: app/views/listings/new.html.haml (hunk 6 above)
Comment (tochman) [reply app/views/listings/new.html.haml]: The `partials` folder is by convention used for partials that relate to the application. If you have code that you want to extract to a partial and that is coupled to a resource, that partial should be placed in the corresponding folder in my opinion (in this case the `listings` folder)

File: app/controllers/profiles_controller.rb [hunk 8]
```diff
@@ -1,4 +1,23 @@
 class ProfilesController < ApplicationController
+  
   def new
+    @profile = Profile.new
   end
+
+  def create
```
Comment (tochman) [inline app/controllers/profiles_controller.rb:10]: The way you do this opens this action to non-authenticated users.

Comment (tochman) [reply app/controllers/profiles_controller.rb:10]: I would look into using a `before_action` and make sure the user is authenticated first. You might argue that you regulate this on the view, but IMO it's not enough.

File: features/user_can_create_profile.feature [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  In order to be able to apply for becoming a cat host
+  I need to create a host profile
+
+  Background:
+    Given the following user exist
+      | email        | password   |
+      | boa@snake.se | pswrd12345 |
+
+  Scenario: Registered user tries to create a Host Profile while not logged in (relatively happy path)
+    Given I visit the landing page
```
Comment (tochman) [inline features/user_can_create_profile.feature:13]: Okay, this is good. But what if the user manually navigates to the page where he can create a new profile. For instance: `http://localhost:3000/profiles/new` ?

File: features/user_can_create_profile.feature [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      | email              | password |
+      | george@saloniki.gr | password |
+    Given the following profile exist for a user
+      | name   | location | description | picture                        | user                |
+      | George | Greece   | Sunny       | https://www.pivotaltracker.com | george@saloniki.gr  |
+    Given I am logged in as "george@saloniki.gr"
+    When I visit the landing page
+    Then I should not see "Become a host"
+    And I should see "New listing"
+    And I should see "Log out"
```
Comment (tochman) [inline features/user_can_create_profile.feature:58]: Same thing here.  Is the current implementation actually hindering the user to create another profile or is it just hiding the link?

File: features/user_can_create_profile.feature (hunk 9 above)
Comment (zanenkn) [reply features/user_can_create_profile.feature:13]: Yes, we raised this with Faraz before and he told us not to worry about it just yet. But ok, we shall look at it then.

File: features/user_can_create_profile.feature (hunk 10 above)
Comment (zanenkn) [reply features/user_can_create_profile.feature:58]: same as above.

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  describe 'Factory' do
+    it 'should have a valid Factory' do
+      expect(FactoryBot.create(:user)).to be_valid
+    end
+  end
+
+  describe 'Association' do
+    it { should belong_to(:profile).optional }
```
Comment (tochman) [inline spec/models/user_spec.rb]: Please use the expect syntax

File: app/models/profile.rb (hunk 4 above)
Comment (zanenkn) [reply app/models/profile.rb]: Ok, done.

File: app/models/user.rb (hunk 5 above)
Comment (zanenkn) [reply app/models/user.rb]: Resolved

File: app/views/listings/new.html.haml (hunk 6 above)
Comment (zanenkn) [reply app/views/listings/new.html.haml]: Will do.

File: app/controllers/profiles_controller.rb (hunk 8 above)
Comment (zanenkn) [reply app/controllers/profiles_controller.rb:10]: done.

File: features/user_can_create_profile.feature (hunk 9 above)
Comment (zanenkn) [reply features/user_can_create_profile.feature:13]: Done.

File: spec/models/user_spec.rb (hunk 11 above)
Comment (zanenkn) [reply spec/models/user_spec.rb]: done.

File: app/views/listings/new.html.haml (hunk 6 above)
Comment (raptorf1) [reply app/views/listings/new.html.haml]: done

**Pull request merged**